vim: Fix end of paragraph deletion when there's no blank lines

dinocosta opened the pull request:
This Pull Request attempts to fix an issue where using `d}` in vim mode would not delete all characters in case there's no blank lines at the end of the buffer.

When calculating the end point for this motion, if there's no blank lines at the end of the buffer, Zed was calculating it to be the last character in the last line. However, if there's a newline at the end of the buffer, it calculates the end point to be the point at the right of the last character.

Here's an example, for the following buffer contents:

```
Hello!
Hello!
```

If the `d}` command is run at `(0, 0)`, the end point will be set to `(1, 5)`. However, fi the same command is run for this buffer instead:

```
Hello!
Hello!

```

The end point will be set to `(1, 6)`, there's a 1 unit difference in the column, which leads to all characters actually being deleted.

Closes #29393 

Release Notes:

- Fixed deleting to the end of paragraph when there's no blank lines

dinocosta committed:
fix(delete): fix end of paragraph deletion for buffers with no newlines

This commit attempts to fix an issue where using `d}` in vim mode would
not delete all characters in case there's no newlines at the end of the
buffer.

When calculating the end point for this motion, if there's no newline at
the end of the buffer, Zed was calculating it to be the last character
in the last line. However, if there's a newline at the end of the
buffer, it correctly calculates the end point to be the character at the
right of the last character.

Here's an example, for the following buffer contents:

```
Hello!
Hello!
```

If the `d}` command is run at (0, 0), the end point will be set to (1,
5). However, fi the same command is run for this buffer instead:

```
Hello!
Hello!

```

The end point will be set to (1, 6), there's a 1 unit difference in the
column, which leads to all characters actually being deleted.

dinocosta commented on crates/vim/src/motion.rs:
```diff
@@ -904,10 +904,22 @@ impl Motion {
-                map.clip_at_line_end(movement::end_of_paragraph(map, point, times)),
-                SelectionGoal::None,
-            ),
+            EndOfParagraph => {
+                let mut display_point =
+                    map.clip_at_line_end(movement::end_of_paragraph(map, point, times));
+
+                // If the point's column is not `0`, that means that the point is not at the beginning of a line.
+                // This can happen, for example, when using the `d}` vim command
+                // on a buffer with no newlines at the end. In that case, we
+                // want to extend the point's column 1 position to the right,
+                // ensuring that all characters are deleted.
+                if display_point.column() != 0 {
+                    display_point =
+                        DisplayPoint::new(display_point.row(), display_point.column() + 1);
+                }
+
+                (display_point, SelectionGoal::None)
+            }
```
@ConradIrwin I'm not sure if this the best place to actually fix this issue, so I'd love your feedback here 🔍

I believe there's multiple places where this could be handled:

- `vim::motion::Motion::range` - It seems there's already some cases handled in that method so I tried to avoid adding it here to avoid making that method more complex
- `vim::motion::Motion::move_point` - This is where I added the fix, as this seems to be the function where Zed actually determined the end point for the motion.
- `editor::movement::end_of_paragraph` - I pondered adding the fix here, but since this issue seems to be Vim-specific it kind of feels like this function should not be touched. As if there's actually only two lines in the buffer, the end of paragraph actually makes sense to be the point for the last character in the last line?

Let me know what you think! If you prefer, I can also schedule some pairing time, thanks! 🙇

ConradIrwin replied:
Hmm.. I think `range` is probably the right fix, as using `}` to go to the end seems to work alright today.

That said, the tests will probably catch mistakes, so where-ever is ok :D.

dinocosta commented:
Here's a quick screen capture showing the fix in action 📹 

https://github.com/user-attachments/assets/057ac592-5105-4b36-9c17-3e3e9c5e468d

dinocosta committed:
refactor(vim): move d} fix to range method

Update the fix for the `d}` delete motion from
`Motion::move_point` to `Motion::range` in order to avoid
breaking other commands that rely on the `}` motion.

This also fixes the tests that were failing and introduces
a new test to ensure the `}` delete motion works as expected.

ConradIrwin committed:
Don't use +1

ConradIrwin commented:
Looks great, thank you!

`+ 1` is not safe in Zed, as the columns are "number of utf-8 encoded bytes since the start of the line", and if you have a multibyte character in the last position, this will panic (because rust is very strict about ensuring strings are valid utf-8). I fixed this by using `movement::saturating_right`, but we could also have used `clip_point` to move the new index to a valid utf-8 offset.

dinocosta commented:
> Looks great, thank you!
> 
> `+ 1` is not safe in Zed, as the columns are "number of utf-8 encoded bytes since the start of the line", and if you have a multibyte character in the last position, this will panic (because rust is very strict about ensuring strings are valid utf-8). I fixed this by using `movement::saturating_right`, but we could also have used `clip_point` to move the new index to a valid utf-8 offset.

Thank you for the detailed explanation! Should have noticed the usage of `movement::saturating_right` right below the code I touched, my bad 🙇
